packages/core/src/api/app/App.tsx

```diff
@@ -130,22 +132,56 @@ export class AppBuilder {
     let rendered = (
       <Switch>
         {routes}
-        <Route
-          render={(props) => (
-            <ErrorPage {...props} status="404" statusMessage="PAGE NOT FOUND" />
-          )}
-        />
+        <Route component={NotFoundErrorPage} />
       </Switch>
     );
 
     if (this.apis) {
       rendered = <ApiProvider apis={this.apis} children={rendered} />;
     }
 
-    return () => <AppContextProvider app={app} children={rendered} />;
+    return () => rendered;
+  }
+
+  getProvider(): ComponentType<{}> {
+    const Provider: FC<{}> = ({ children }) => (
+      <AppContextProvider app={this}>{children}</AppContextProvider>
+    );
+    return Provider;
+  }
+
+  verify() {
+    const pluginIds = new Set<string>();
+
+    for (const plugin of this.plugins) {
+      const id = plugin.getId();
+      if (pluginIds.has(id)) {
+        throw new Error(`Duplicate plugin found '${id}'`);
+      }
+      pluginIds.add(id);
+    }
   }
 }
 
-export function createApp() {
-  return new AppBuilder();
+/**
+ * Creates a new Backstage App.
+ */
+export function createApp(options?: AppOptions) {
+  const DefaultNotFoundPage = () => (
+    <ErrorPage status="404" statusMessage="PAGE NOT FOUND" />
+  );
+
+  const apis = options?.apis ?? ApiRegistry.from([]);
+  const icons = { ...defaultSystemIcons, ...options?.icons };
+  const plugins = options?.plugins ?? [];
+  const components = {
+    NotFoundErrorPage: DefaultNotFoundPage,
+    ...options?.components,
+  };
+
+  const app = new AppImpl({ apis, icons, plugins, components });
+
+  app.verify();
```

Comment:
should the verification be done before we instantiate an AppImpl?

Reply:
I think it's nicer to have the verification step inside AppImpl, createApp is really just a wrapper to provide default values

Reply:
yeah. was just wondering if we should break even earlier if the app has duplicate plugin ids. Like doing new AppImpl is going to go through all the plugin outputs, construct all the routes to be rendered and then we verify. I am understanding this right?

Reply:
nvm. got it.

Reply:
i mean the constructor could have done the verification directly, but sure